### DISCUSSION TITLE
433MHz transmitter and receiver

### DISCUSSION THREAD

**Discussion by kenwawa01** [merbanan/rtl_433#2941] [Q&A]
Hi!
By default in rtl_433 project we use receivers to analyse TPMS data. I was wondering how complicated it would be also to generate messages to be sent to tpms module. Something like tpms sensor(s) simulator. 

Which device could be then used for receiving/transmitting data? Which one would you recommend?

Do you know any project which could be some kind of example for transmitting tpms data (simulating tpms sensors)?

All links would be welcomed.

**Comment by zuckschwerdt** [comment]
You'd need hardware capable of sending FSK. Could be some SDR with TX or dedicated tranceivers like a CC1101.

**Comment by kenwawa01** [comment]
Thanks for answering. I'm just entering 433MHz wold thus my questions might be very basic.

Could you give few example of those SDRs with TX you would recommend to work with rtl_433 project?

Does any DVB tuners have the possibility to "produce signal"? Sth like: https://pl.aliexpress.com/item/1005006420997014.html I'm using it for recording 433MHz and it works surprisingly well.

What about SDRs like this https://pl.aliexpress.com/item/1005005952566458.html ?

With CC1101, I've never seen it in USB version to be connected to laptop. Sure you can use usb-rs232 interface but it will not be very stable "mechanically". I mean it might not be convenient to be use longer with laptop unless you will put everything in nice box and solder permanently.

Will rtl_433 work without problems with CC1101 connected via usb-rs232 interface?

**Reply by zuckschwerdt** [reply]
rtl_433 has no TX capability. The RTL-SDR you linked do not have TX, no "tuners" have that. You get TX with SDRs in a higher price range. E.g. LimeSDR, PlutoSDR, HackRF.

The CC1101 (and other such transceivers) need I2C (not UART serial) and good programming skills.

**Comment by kenwawa01** [comment]
Sure I know that rtl_433 has no TX capability. But when buying I would like to buy sth which will work with rtl_433 in read mode and later I could try to find some other projects (plus perhaps modify them) to use the same device in write mode.

Do you perhaps know any fork rtl_433 projects which work with CC1101?

**Reply by zuckschwerdt** [reply]
[rtl_433 esp32](https://github.com/NorthernMan54/rtl_433_ESP) does.

**Reply by kenwawa01** [reply]
> [rtl_433 esp32](https://github.com/NorthernMan54/rtl_433_ESP) does.

I'm wondering if it would be difficult to make it working on normal PC laptop with Linux... I will look on this project as there is chance I could use sth. Thanks a lot.

**Reply by zuckschwerdt** [reply]
An ESP32, even the obsolete low-end esp8266 can be the adapter for that. You will need to program it for that application though.

**Reply by kenwawa01** [reply]
> An ESP32, even the obsolete low-end esp8266 can be the adapter for that. You will need to program it for that application though.

I've found sth like this https://pl.aliexpress.com/item/1005006706464169.html They are writing that they use internal usb-serial port connection (some kind of serial virtual port). I'm considering buying this and checking it it works with rtl_433. Any guess?

**Reply by zuckschwerdt** [reply]
The CC1101 is not an SDR and won't work with rtl_433. You will need to program your own application.

**Reply by kenwawa01** [reply]
Sure, thanks a lot for help. Knowing well rtl_433, could you say which .c files should be reprogrammed/changed? On which I should focus? It will help me a lot.`

**Reply by zuckschwerdt** [reply]
Start by reading the spec pdf of the CC1101 or maybe read about the SX127x (SX1276/77/78/79) -- there are ready built modules with that transceiver: https://www.lilygo.cc/products/lora3

**Reply by kenwawa01** [reply]
Thank you for your suggestions. I've looked a bit more on CC1101 and it doesn't seem to be very complicated to use it although people are writing that for Arduino they have problems with sensors reading (incorrect parameters?), ex here: https://forum.arduino.cc/t/cc1101-and-tpms-sensors/1121534 Sure I would prefer to use it with laptop computer with Linux so it might be a bit challenging with libraries...

Btw, will rtl_433 for reading tpms messages (no TX) work with DVB-T2 usb tuners based on RTL2832P (the last character is different), ex. sth like that: https://pl.aliexpress.com/item/1005003105710626.html

**Reply by pswiatki** [reply]
> Btw, will rtl_433 for reading tpms messages (no TX) work with DVB-T2 usb tuners based on RTL2832P (the last character is different), ex. sth like that: https://pl.aliexpress.com/item/1005003105710626.html

If this is a question - then yes it will. NOTE: I would strongly recommend using dongles that are considered Proper, see: https://www.rtl-sdr.com/buy-rtl-sdr-dvb-t-dongles/ as those are much more stable than some cheap ones ("clones").

**Reply by kenwawa01** [reply]
Yes, it was the question. :P Thanks for this hint. :) For simple use those dvb-t/t2 should be enough, at least i hope, and u can always use them for watching tv. Those rtl-sdr v3/v4 are not for tv watching right?

**Reply by pswiatki** [reply]
Dear @kenwawa01 V3/V4 contain exactly the same chip RTL2832, which does DVB-T demodulation and decoding. Note: a different driver is required on the host system to receive and decode digital, terrestrial TV. Hope this helps.

**Reply by kenwawa01** [reply]
But are you sure that it's only the question of this main chip? I've done some small searching and I found many suggestion that v3/v4 are enough for dvb-t but not good enough for dvb-t2 and it will not work. Sure I didn't check it by myself.

**Reply by pswiatki** [reply]
Don't know anything about DVB-T2. Nevertheless, this is outside of the above topic (`rtl_433`), so I can care less about DVB-T support. The important thing for RTL-SDR is that the dongles can act as cheap and effective RF samplers with tuners capable of receiving signals from a wide range of frequencies (the newer ones can also do direct sampling of HF). This is enough to know about those devices.

**Comment by kenwawa01** [comment]
Ok I bought device with RTL2832P https://pl.aliexpress.com/item/1005001812444617.html and it works much worse than the one with RTL2832U https://pl.aliexpress.com/item/1005006385049099.html

This P is loosing many messages which were read without problems with U. Moreover with P I have quite often messages which I didn't notice in case of U version.

` bit_buffer_add_bit: Warning: row count limit (50 rows) reached`

I tried using `-g=-5` or `-g=5` or similar values but it doesn't help. I tried also` -Y autolevel -M level -M noise` but I'm not able to improve scanning.

The only thing which helps a bit is to reduce the antenna length but it still gives worse results than in case of U version.

Is that more antenna problem or this new device has much worse radio reception or might it be that rtl_433 code is missing sth for this new version, some kind of automatic signal leveling etc.?

**Comment by zuckschwerdt** [comment]
`bit_buffer_add_bit: Warning: row count limit (50 rows) reached` is a sign of noise that looks like data. You could compare both receivers using a SDR UI, or grab samples with both (or record for a short time) and compare in https://triq.org/pdv/

**Comment by kenwawa01** [comment]
It might be that P is more noisy. But with P I'm missing more or less 50% of correct messages which I was able to catch without problems with U. Thus it suggest that P might be so noisy that it cannot read the correct messages. Ex. if tpms sensor repeat the same message 4 times, I get ex. repetition 1 and repetition 3 only. Wile with U nothing missed. Sure in the same conditions.

I can't compare two in the more detailed way as you suggest, as U was borrowed and I don't have it any longer. I bought P for myself and it seems it was not the best idea.

I'm only not sure if it's purely hardware problem or P might have sth slightly different in software part and it requires ex. some additional parameters when being used...?

**Reply by zuckschwerdt** [reply]
Interesting, thanks! Take a look at a sample or recording anyway, you might notice something in the spectrogram, could be interference or such?
